### PR TITLE
router: only cache single route for the active request

### PR DESCRIPTION
**Pull request by wbpcode**
Commit Message: router: only cache single route for the active request
Additional Description:

In the previous impl of the route table, every active request will keep a reference of the whole route table. If there route table is updated frequently and there are some long-live http request, the old versions route tables cannot be released timely and increase the overhead of memory.

This PR try to fix this problem by cache single route for active request.

See #20800 for more detailed background.

Risk Level: High. Core code patch. Change the lifetime of route table.
Testing: Added/fixed.
Docs Changes: n/a.
Release Notes: n/a.
Platform Specific Features: n/a.

### CODE REVIEW

File: envoy/router/cluster_specifier_plugin.h
Comment (wbpcode) [inline envoy/router/cluster_specifier_plugin.h:29]: comment: by this way, we the cluster specifier plugin could take the ownership of the parent route.

File: envoy/router/router.h
Comment (wbpcode) [inline envoy/router/router.h:613]: comment: split the interfaces that needn't depend on router matcher out to the `SharedConfig`. Then every virtualhost could keep a reference to the `SharedConfig`.
When the entire route config is destroyed, the part of `SharedConfig` will still alive until all active requests are completed.

File: source/common/http/conn_manager_impl.cc
Comment (wbpcode) [inline source/common/http/conn_manager_impl.cc:1798]: comment: Because we will not cache the entire route config in the active request, we need get the route config from the hcm again when we refresh the route.
This will result in a minor behavior change. That's say if a request is comming and a initial route is selected. And then, the route config is updated by the RDS. And then the cached route is clear and refreshed by one filter in the filter chain. In the previous impl, the old version route config will be used to refresh the route and in the new impl the latest route config will be used to refresh the route.

**Comment by alyssawilk** [comment]
/wait on CI -  I think the router drop is real
Or were you waiting for first pass before writing tests?  If so  I think @adisuissa could do a round as this looks like it's a config lifetime change

**Comment by JakeCooper** [comment]
Hello from the Railway folks (original filers of #20800)

Patch is looking reasonably good! Will circle back if anything changes but, hopefully that provides a nice little datapoint "in the wild"

CleanShot 2023-03-14 at 17 21 32@2x

**Comment by wbpcode** [comment]
Will to write some tests to improve the coverage of the `common/router`.

It would be great if more experts could help to review this PR. This lifetime change always be dangerous patch.

**Comment by wbpcode** [comment]
cc @JakeCooper Thanks for your feedback. Note this patch is no completed and it couldn't be used in the production. Especially if scoped rds & ondemand rds are used by you.

**Comment by wbpcode** [comment]
I rechecked the code of ondemand rds. Seems the ondemand callback is managed by the route config provider. So this patch will not effect it.

**Review by adisuissa — COMMENTED** [review]
Thanks! Overall, the code changes LGTM, but would add a few tests that validate that the new behavior is maintained (if the config is changed during execution time, the new one is latched onto).

IMO there is a fundamental question here: when should the filter-manager latch to the route table (see [Matt's comment](https://github.com/envoyproxy/envoy/issues/20800#issuecomment-1464151544)?
Prior to this PR the routing table is latched during the `decodeHeaders()` call.
This PR changes it to latch later when `route()` is called.
Changing the routing table view in the midst of routing is possible, but it should be reviewed by someone that understands the details of the router.
Also cc @envoyproxy/senior-maintainers to chime in on this.

Another alternative to consider is to explicitly release the routing table, once a request "realizes" that the route is fixed (i.e., no need to reroute). It may be a bit easier to understand the code, and clearly find the cases where there could be an old-new view conflict of the routing table. WDYT?

**Comment by wbpcode** [comment]
> Prior to this PR the routing table is latched during the decodeHeaders() call. This PR changes it to latch later when route() is called.

In fact, we still will get the route table in the `decodeHeaders()` to get the initial route.

> Another alternative to consider is to explicitly release the routing table, once a request "realizes" that the route is fixed (i.e., no need to reroute). It may be a bit easier to understand the code, and clearly find the cases where there could be an old-new view conflict of the routing table. WDYT

Sounds a good idea. Then we can keep current behaviors completely. Typically, we can do this in the `router` filter. And I think we can also make this a interface of `StreamFilterCallback`. May be call it `freezeRouteCache` or something. 
cc @mattklein123 @alyssawilk

File: source/common/http/conn_manager_impl.cc
Comment (wbpcode) [reply source/common/http/conn_manager_impl.cc:1809]: The target services/clusters is managed by the cm and I think the route change will not effects it.

Except the behavior change, this refresh should be safe.

But anyway, I think your idea to clear the cached route table explicitly is great, so I will update this patch in next days.

**Comment by wbpcode** [comment]
cc @adisuissa Hi, could you take a look again when you have time?

Now the PR is cleaner. The PR contains two parts:
1. Refactor the route config to make the route/vh could keep the ownership of shared part of route config. By this way, we can release route config and keep only single route for the active requests.
2. Clear snapped route config in the active stream after the response headers have been processed. For the long-live HTTP request, although the entire request will has a long lifetime, the response headers should be returned and be handled in a short time (?).

File: source/common/http/conn_manager_impl.cc
Comment (adisuissa) [inline source/common/http/conn_manager_impl.cc:1441]: My knowledge on when `refereshCachedRoute()` is called is limited, but if there's an expectation that the function will change the internal state (calling `setRoute()`), then the added code may break this.
I suggest adding a comment explaining the reasoning behind this.

File: source/common/http/conn_manager_impl.cc
Comment (adisuissa) [inline source/common/http/conn_manager_impl.cc:1601]: Please add a comment explaining why it is ok to block the route-cache at this stage.

File: source/common/http/conn_manager_impl.cc
Comment (adisuissa) [inline source/common/http/conn_manager_impl.cc:1860]: OOC, if the filters had a shared_ptr to the route, would this still be necessary?

**Review by adisuissa — COMMENTED** [review]
Thanks. Overall I think it is easier to understand this change, but would need someone with more understanding of routing to think about the impact of this change.
It may be beneficial to have a runtime guard against this (I think it modifies current behavior).

File: source/common/http/conn_manager_impl.cc
Comment (wbpcode) [reply source/common/http/conn_manager_impl.cc:1441]: get it.

File: source/common/http/conn_manager_impl.cc
Comment (wbpcode) [reply source/common/http/conn_manager_impl.cc:1601]: get it.

File: source/common/http/conn_manager_impl.cc
Comment (wbpcode) [reply source/common/http/conn_manager_impl.cc:1860]: If the filters cached the shared_ptr to the route, then this is not necessary. However, there is no any guard for that.
Most filter think all the route config will be cached by the active stream and will not destoryed until the request is completed.

**Comment by wbpcode** [comment]
> but would need someone with more understanding of routing to think about the impact of this change.

This patch change some core code. After the first-pass, it still should be reviewed by at least one of @alyssawilk and @mattklein123. 😺 

> It may be beneficial to have a runtime guard against this (I think it modifies current behavior).

We will block the route-cache after response headers processing. After that time, any route-cache refreshing should be meaningless. And before the route-cache is blocked, no any behavior will be impacted.
 So I am not sure whether its necessary to add this runtime guard to guard these meaningless operations. cc @envoyproxy/runtime-guard-changes

File: source/common/router/config_impl.h
Comment (mattklein123) [inline source/common/router/config_impl.h:1184]: Same question.

File: envoy/router/router.h
Comment (mattklein123) [inline envoy/router/router.h:640]: Shouldn't this return shared_ptr to avoid dangling lifetimes? See all my other comments about circular references though.

File: source/common/http/conn_manager_impl.h
Comment (mattklein123) [inline source/common/http/conn_manager_impl.h:456]: When I first read this, I thought this was somehow to avoid dangling lifetimes where there are raw references inside the route structure, and I was going to suggest figuring out a way to use shared/weak to avoid this.

But now I see that you do use shared_ptr inside the route table so I'm confused:
1) Why is this necessary? Aren't all references now strong and safe?
2) How do we not have circular references and memory leaks inside the route table?

It seems like we want to move to handing out shared_ptr references where everything is strongly held.

The issue is how to avoid circular references. Maybe I'm missing something, but it seems like the only way for this to work in practice is for all "up links" inside the route table to be weak_ptr, and when a route is handed out, we walk the links and lock them as shared. 

This would then allow everything to be dropped around the reference.

Is there a better way? Am I missing something?

**Review by mattklein123 — COMMENTED** [review]
Thanks for working on this. Some high level comments to get started.

File: source/common/http/conn_manager_impl.cc
Comment (wbpcode) [reply source/common/http/conn_manager_impl.cc]: The `traversePerFilterConfig` or `mostSpecificPerFilterConfig` will not return shared_ptr but raw pointer or raw reference.

I think `header_to_metadata`, `header_mutation`, etc will store raw pointer to the route level filter config in the decoding phase and use it in the encoding phase. And I think there should be lots third party filters will do this like that.

It's easy to update the filters in our repo, but I am afraid this will breakdown lots third party filters. And this problem is hard to be found in the compile time.

**Comment by wbpcode** [comment]
I think @JakeCooper 's test had ensured that there is no circular reference, or the memory usage will never decrease.

Here is the point.

Although the VH need a reference to the global route config. But note that the VH needn't access everything of the route config. Only the part shared by all VHs is necessary for the VH. So the split the route config to two parts:
- One shared part that shared by all VHs.
- A VH list.

Every VH will keep a shared pointer to the shared part of route config, not the shared pointer to the whole route config. So, there is no circular reference.
Similarly, I will split VH to shared part and route list. Every route will keep a shared pointer to the shared part of VH, not the shared pointer to the whole VH.

memory-route-config

When the route config is destroyed, the route cached by the active stream, the shared part of the route's parent VH, the shared part of route config, all these configration will still be alive until the request is completed.

Although the weak_ptr is possible solution. However, after the route config is destroyed, some necessary config that store in the VH or route config will be invalid.

File: source/common/router/config_impl.h
Comment (wbpcode) [reply source/common/router/config_impl.h:329]: See https://github.com/envoyproxy/envoy/pull/26045#issuecomment-1478844902

File: source/common/router/config_impl.h
Comment (wbpcode) [reply source/common/router/config_impl.h:1184]: See https://github.com/envoyproxy/envoy/pull/26045#issuecomment-1478844902

File: envoy/router/router.h
Comment (wbpcode) [reply envoy/router/router.h:640]: I think it's unnecessary because the `SharedConfig` will always be valid for the VH.

File: source/common/router/config_impl.h
Comment (wbpcode) [reply source/common/router/config_impl.h:971]: When we create a instance of `DynamicRouteEntry`, the instance will returned directly to the active request. The instance will not be stored by the `RouteEntryImplBase`, so there is no circular reference.

The special case is the `WeightedClusterEntry` which is derived from `DynamicRouteEntry` but the instances of `WeightedClusterEntry` will be store in the `RouteEntryImplBase`. So we will keep the `owner_` of `WeightedClusterEntry` be nullptr to avoid circular references.

Comment (mattklein123) [reply source/common/router/config_impl.h:971]: OK I see, thanks. Can you add more comments around all of this? It's kind of confusing to look through. The other ones about the shared part make sense. Thanks for the explanation.

File: source/common/http/conn_manager_impl.cc
Comment (mattklein123) [reply source/common/http/conn_manager_impl.cc]: OK I see. In this particular case I would argue that we should change the type name to force a compile error and then have a release note about how to fix it. I'm not sure how large of a change that would be though. Perhaps by just changing the name of the type alias it would be enough?

Usually I would not argue for such a thing, but it seems not great to add this type of extra complexity when we have a way to force people to take notice and fix their code. WDYT?

File: source/common/http/conn_manager_impl.h
Comment (wbpcode) [reply source/common/http/conn_manager_impl.h:456]: These cleared route entries is stored by the active stream to ensure all the config used by the filter chain will be valid in the whole lifetime of the request.

For example, at the decoding phase, filter A get a route level filter config from the route and store the raw pointer of the config in the filter instance (lots of filters do this). Then following filter B clear and refresh the route.
At the encoding phase, the filter A may access the route level filter config again by the raw pointer (header to metadata, header mutation, and maybe lots of thrid party filters do this). If the the active stream doesn't keep the cleared route entries, the pointer to the route level config may be invalid because the RDS updates.

Comment (mattklein123) [reply source/common/http/conn_manager_impl.h:456]: Yup understood, see my other comment about potentially forcing a compiler error so we make people fix their code.

**Comment by wbpcode** [comment]
To solve #20800, there are two key points:
- 1. fine-grained lifetime management of route config. Different routes/VHs of same route config could have different lifetime, to ensure it's possible to cache only one single route for active stream.
- 2. clear snapped route config timely and ensure there is no dangling reference or pointer in the active stream (mainly in the HTTP filter chain).


I split the route config and VH to different parts to land the first point. I keep all the cleared or refreshed route entries in the active stream to land the second point.

File: source/common/http/conn_manager_impl.cc
Comment (wbpcode) [reply source/common/http/conn_manager_impl.cc]: Do you mean change the basic type name to notify the third party developers that their filters are broken.

And then to tell them that they should keep a copy of route (or some others way) in the filter if they want to keep route level filter config's pointer in the filter by a release note?

File: source/common/router/config_impl.h
Comment (wbpcode) [reply source/common/router/config_impl.h:971]: get it.

File: source/common/http/conn_manager_impl.cc
Comment (mattklein123) [reply source/common/http/conn_manager_impl.cc]: I'm saying to change the name of the fewest number of things that will cause all users of storing a route to have a compile break. We can fix all filters/extensions in the public repo, and then have a release note for other users to store a strong pointer. I think this is worth it in this case vs. carrying the additional lifetime hacks in the HCM. cc @alyssawilk for a second opinion on this.

Comment (wbpcode) [reply source/common/http/conn_manager_impl.cc]: Sounds ok to me. But I think the things need to change wouldn't be 'few'.

we can update the return type of 'perFilterConfig' and the method name to get route.
Then almost all third party filters that try to store the route level config or route will get compile error.

Comment (mattklein123) [reply source/common/http/conn_manager_impl.cc]: Yeah I'm not sure what would be the "easiest" thing to do here, but I do think this is the best path forward. Do you mind prototyping it to see how bad it is?

Comment (wbpcode) [reply source/common/http/conn_manager_impl.cc]: I can have a try in a separated pr first to see how much code we need to change.

**Comment by mattklein123** [comment]
Sorry I will look soon.

File: changelogs/1.25.0.yaml
Comment (mattklein123) [inline changelogs/1.25.0.yaml:79]:
```suggestion
    prohibit route refresh after the response headers have been sent,  and clear the cached route table to more aggressively free memory. This behavior can be reverted by setting runtime flag ``envoy.reloadable_features.prohibit_route_refresh_after_response_headers_sent`` to false.
```

File: source/common/http/conn_manager_impl.h
Comment (mattklein123) [inline source/common/http/conn_manager_impl.h:479]: Just pointing out that if some filter does somehow use the extended route table after the response has been sent (logging callback, onDestroy, who knows), this is still going to crash. But I guess we can see?

**Review by mattklein123 — CHANGES_REQUESTED** [review]
LGTM with a few comments, thank you for tackling this tough problem! Awesome.

/wait

File: source/common/http/conn_manager_impl.h
Comment (wbpcode) [reply source/common/http/conn_manager_impl.h:316]: Yeah, you are right. I think an ENVOY_BUG we be helpful. Thanks.
This flag is only checked when we try to clear or fresh the route cache. But after the response headers is sent (at that time the flag will be set to true), that should never be happened.

File: source/common/http/conn_manager_impl.h
Comment (wbpcode) [reply source/common/http/conn_manager_impl.h:479]: What do you mean by the `extended route table`? The `Router::ConfigImpl` instance? I think the filter should never and cann't access the whole route table? 🤔

Comment (mattklein123) [reply source/common/http/conn_manager_impl.h:483]: Sorry I just meant that it's possible for a filter to use route data after response headers have been sent (logging, onDestroy, etc.). But I guess now thinking about it, anything that a filter could access via raw pointer is kept alive via the hack, right? So there is no chance of crash in any further use?

**Review by mattklein123 — DISMISSED** [review]
Awesome work. LGTM!

**Comment by wbpcode** [comment]
Hi, could you give a stamp again? 😸 @ Conflict resolved and minor update to the changelog format.

**Review by mattklein123 — APPROVED** [review]

**Pull request merged**
